### PR TITLE
feat(event_handler): use custom serializer during openapi serialization

### LINKED ISSUE #3892: Bug: Mongo ObjectIds not serializing out of the box
### Expected Behaviour

My app has a custom serializer. I used enable_validation. A dict with Mongo ObjectId should have been serialized.

```
from bson import json_util
...
def custom_serializer(obj) -> str:
    ...
    """Your custom serializer function ApiGatewayResolver will use"""
    return json_util.dumps(obj)

app = APIGatewayRestResolver(serializer=custom_serializer, enable_validation=True)
```

### Current Behaviour

If we use enable_validation, the openapi modules uses its own encoders which cannot handle ObjectIds. 

```
Traceback (most recent call last):
File "/var/task/aws_lambda_powertools/event_handler/openapi/encoders.py", line 265, in _dump_other
data = dict(obj)
TypeError: 'ObjectId' object is not iterable
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
File "/var/task/aws_lambda_powertools/event_handler/openapi/encoders.py", line 269, in _dump_other
data = vars(obj)
TypeError: vars() argument must have __dict__ attribute
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
File "/var/task/aws_lambda_powertools/event_handler/api_gateway.py", line 1985, in _call_route
route(router_middlewares=self._router_middlewares, app=self, route_arguments=route_arguments),
File "/var/task/aws_lambda_powertools/event_handler/api_gateway.py", line 400, in __call__
return self._middleware_stack(app)
File "/var/task/aws_lambda_powertools/event_handler/api_gateway.py", line 1291, in __call__
return self.current_middleware(app, self.next_middleware)
File "/var/task/aws_lambda_powertools/event_handler/middlewares/base.py", line 121, in __call__
return self.handler(app, next_middleware)
File "/var/task/aws_lambda_powertools/event_handler/middlewares/openapi_validation.py", line 121, in handler
return self._handle_response(route=route, response=response)
File "/var/task/aws_lambda_powertools/event_handler/middlewares/openapi_validation.py", line 128, in _handle_response
response.body = self._serialize_response(
File "/var/task/aws_lambda_powertools/event_handler/middlewares/openapi_validation.py", line 187, in _serialize_response
return jsonable_encoder(response_content)
File "/var/task/aws_lambda_powertools/event_handler/openapi/encoders.py", line 108, in jsonable_encoder
return _dump_dict(
File "/var/task/aws_lambda_powertools/event_handler/openapi/encoders.py", line 212, in _dump_dict
encoded_value = jsonable_encoder(
File "/var/task/aws_lambda_powertools/event_handler/openapi/encoders.py", line 138, in jsonable_encoder
return _dump_other(
File "/var/task/aws_lambda_powertools/event_handler/openapi/encoders.py", line 272, in _dump_other
raise ValueError(errors) from e
ValueError: [TypeError("'ObjectId' object is not iterable"), TypeError('vars() argument must have __dict__ attribute')]
```

### Code snippet

```python
# Any object which contains a value of type Mongo ObjectId

ObjectId('666f6f2d6261722d71757578')
```

### Possible Solution

My workaround is to monkey patch openapi encoder. Ideally openapi should reuse json encoder of APIGatewayRestResolver, or allow plugging in our own function.

```
from bson import ObjectId
from aws_lambda_powertools.event_handler.openapi.encoders import ENCODERS_BY_TYPE
ENCODERS_BY_TYPE[ObjectId] = lambda o: str(o)
```

### Steps to Reproduce

Serialize an object which contains Mongo ObjectId

```
ObjectId('666f6f2d6261722d71757578')
```

### Powertools for AWS Lambda (Python) version

latest

### AWS Lambda function runtime

3.8

### Packaging format used

PyPi

### Debugging logs

_No response_

### PR DESCRIPTION
**Pull request by rubenfonseca**
**Issue number:** #3892

### Changes

> Please provide a summary of what's being changed

This PR modifies the OpenAPI validation middleware to use the any custom serialized provided by the user on the event handler, when the normal serialization mechanisms fail.

This allows you to use your own custom serializer when returning custom objects
from the handlers.

### User experience

> Please share what the user experience looks like before and after this change

```python
def test_openapi_serialize_other(gw_event):
    # GIVEN a custom serializer
    def serializer(_):
        return "hello world"

    # GIVEN APIGatewayRestResolver is initialized with enable_validation=True and the custom serializer
    app = APIGatewayRestResolver(enable_validation=True, serializer=serializer)

    # GIVEN a custom class
    class CustomClass(object):
        __slots__ = []

    # GIVEN a handler that returns an instance of that class
    @app.get("/my/path")
    def handler():
        return CustomClass()

    # WHEN we invoke the handler
    response = app(gw_event, {})

    # THEN we the custom serializer should be used
    assert response["body"] == "hello world"
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Review by sthulb — DISMISSED** [review]

File: aws_lambda_powertools/event_handler/middlewares/openapi_validation.py [hunk 1]
```diff
@@ -55,6 +55,9 @@ def get_todos(): List[Todo]:
     ```
     """
 
+    def __init__(self, response_serializer: Optional[Callable[[Any], str]] = None):
+        self._response_serializer = response_serializer
```
Comment (leandrodamascena) [inline aws_lambda_powertools/event_handler/middlewares/openapi_validation.py]: Is is possible to add docstring here @rubenfonseca?

File: aws_lambda_powertools/event_handler/openapi/encoders.py [hunk 2]
```diff
@@ -29,6 +29,7 @@ def jsonable_encoder(  # noqa: PLR0911
     exclude_unset: bool = False,
     exclude_defaults: bool = False,
     exclude_none: bool = False,
+    custom_serializer: Optional[Callable[[Any], str]] = None,
```
Comment (leandrodamascena) [inline aws_lambda_powertools/event_handler/openapi/encoders.py:32]: Please add this new parameter in the docstring.

File: aws_lambda_powertools/event_handler/openapi/encoders.py [hunk 3]
```diff
@@ -134,6 +135,10 @@ def jsonable_encoder(  # noqa: PLR0911
         if isinstance(obj, classes_tuple):
             return encoder(obj)
 
+    # Use custom serializer if present
+    if custom_serializer:
+        return custom_serializer(obj)
```
Comment (leandrodamascena) [inline aws_lambda_powertools/event_handler/openapi/encoders.py:142]: I'm wondering here whether a custom serializer should take precedence over any other encoder. This is an explicit desire of the customer who wants to use a custom serializer. In this implementation, we only use the custom serializer if not an instance of Pydantic, Dataclass, str, int, and others.

What do you think? Make sense?

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
Hello @rubenfonseca! I just left some small suggestions to improve code maintenance and I would like to hear your opinion on the question I asked.

Thanks, Ruben! This is a great improvement.

File: aws_lambda_powertools/event_handler/openapi/encoders.py (hunk 3 above)
Comment (rubenfonseca) [reply aws_lambda_powertools/event_handler/openapi/encoders.py:142]: @leandrodamascena please note that this is *not* the response serializer. That is still used if the customer provides one. This is the *validation* serializer. When an object comes out of a function, we need to prepare it for validation with Pydantic, and for that we need to transform it into something that is `jsonable`. Internal mechanisms should still take precendence, but here we just add the option to prepare an unknown object for validation.

Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/openapi/encoders.py:142]: Thanks for the clear explanation, Ruben! Make sense about internal mechanisms and its precedence.

Comment (rubenfonseca) [reply aws_lambda_powertools/event_handler/openapi/encoders.py:142]: Actually your question made me think, and I'm renaming the parameter from `response_serializer` to `validation_serializer`, so it's more clear that it's not about the response.

**Review by leandrodamascena — APPROVED** [review]
APPROVED!!

**Pull request merged**